### ISSUE TITLE
autodiff fails under numpy 1.23

### ISSUE THREAD

**Issue by jakevdp** [google/jax#11221]
Something in the numpy 1.23 release broke autodiff; in particular, float0 arrays are being converted to bool:
```python
import jax
import numpy
print(numpy.__version__)
print(jax.grad(lambda x: x+0., allow_int=True)(1).dtype)
```
Newest numpy:
```
1.23.0
bool
```
previous version:
```
1.22.4
[('float0', 'V')]
```

**Comment by jakevdp** [comment]
Temporary pinning to fix CI in https://github.com/google/jax/pull/11220

**Comment by jakevdp** [comment]
Located the issue to this line: https://github.com/google/jax/blob/b623ed58b0415f31b4c2f2c389220f1e51de8d93/jax/interpreters/ad.py#L65

In numpy 1.22, `dtype(t) is float0` is True, and in numpy 1.23 `dtype(t) is float0` is False.

**Comment by jakevdp** [comment]
This is the operative change in numpy:
```python
import numpy as np
from jax import dtypes
print(np.__version__)
print(np.result_type(dtypes.float0) is dtypes.float0)
```
Before:
```
1.22.4
True
```
After:
```
1.23.0
False
```

**Comment by jakevdp** [comment]
I don't know for sure, but I suspect this is the relevant change: https://github.com/numpy/numpy/pull/19345/

**Comment by jakevdp** [comment]
OK, I think this is the result of the new `ensure_canonical` functionality added in https://github.com/numpy/numpy/pull/19226

Part of the reason I think this is because of this little quirk:
```python
In [1]: from jax.dtypes import float0                                                                                                                                                                                                                                                                            

In [2]: import numpy as np                                                                                                                                                                                                                                                                                       

In [3]: np.result_type(float0) is float0                                                                                                                                                                                                                                                                         
Out[3]: False

In [4]: np.result_type(float0, float0) is float0                                                                                                                                                                                                                                                                 
Out[4]: True
```
Along with the fact that `NPY_DT_CALL_ensure_canonical` is only called on the result for the single-input case: https://github.com/numpy/numpy/blob/89b942e41b45ddd91762e704d3f2ca5677a320b7/numpy/core/src/multiarray/convert_datatype.c#L1633

I'm not sure exactly the mechanism whereby this generates a copy, but it seems clear that the fix could go one of two ways:

1. Somehow change our definition of `float0` so it is in a canonical form that doesn't generate a copy when `NPY_DT_CALL_ensure_canonical` is called, or
2. Stop relying on `is` identity to determine if a dtype is equal to `float0`

Option 2 is probably safer in the long run, as I can find no explicit guarantee that functions like `np.dtype` and `np.result_type` will return no-copy references for non-built-in dtypes. However, equality checks for numpy dtypes are annoyingly polymorphic, so this must be used with care (probably `dtype is float0` should be replaced with `np.dtype(dtype) == float0` for safety). Also, since equality check is slightly more expensive than identity check, and since this is used so often throughout the JAX package, this change may have a trace-time performance impact.

I'm going to look deeper and try to find exactly where the copy is being generated, but unless something else comes to light, I think option 2 is our best bet.

**Comment by jakevdp** [comment]
I believe the `ensure_canonical` check for void types leads to this function: https://github.com/numpy/numpy/blob/d4baf3641471971552206b02f94323a54b54d99a/numpy/core/src/multiarray/dtypemeta.c#L295 which ends up generating a copy for JAX's definition of `float0`. This was changed a few months back in https://github.com/numpy/numpy/pull/19346

Here's the TODO that broke our use-case: https://github.com/numpy/numpy/blame/d4baf3641471971552206b02f94323a54b54d99a/numpy/core/src/multiarray/dtypemeta.c#L324-L325

**Comment by jakevdp** [comment]
Additionally, there are at least two regressions in numpy v1.23 that are causing test failures in the JAX test suite: https://github.com/numpy/numpy/issues/21838 and https://github.com/numpy/numpy/issues/21840

**Comment by jakevdp** [comment]
If anyone encounters this, the fix was part of JAX v0.3.14.
